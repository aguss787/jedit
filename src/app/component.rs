pub mod confirm_dialog;
pub mod loading;
pub mod popup;
pub mod preview;
pub mod scrollbar;
pub mod workspace;

#[cfg(test)]
mod test_render;
