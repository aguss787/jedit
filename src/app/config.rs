use std::{fs::File, io::Read};

use byte_unit::{Byte, Unit};
use serde::Deserialize;

#[derive(Debug)]
#[cfg_attr(test, derive(PartialEq))]
pub struct Config {
    pub max_preview_size: Byte,
}

impl Default for Config {
    fn default() -> Self {
        Self {
            max_preview_size: Byte::from_u64_with_unit(1, Unit::MiB)
                .expect("failed to build default max_preview_size"),
        }
    }
}

impl Config {
    pub fn load() -> Self {
        Self::default().patch_from_files(&["/etc/jedit", "~/.jedit", ".jedit"])
    }

    fn patch_from_files(self, files: &[&str]) -> Self {
        files
            .iter()
            .map(File::open)
            .filter_map(Result::ok)
            .filter_map(|mut file| {
                let mut content = String::new();
                file.read_to_string(&mut content).ok()?;
                Some(content)
            })
            .filter_map(|content| toml::from_str(&content).ok())
            .fold(self, Self::patch)
    }

    fn patch(mut self, patch: ConfigPatch) -> Self {
        if let Some(max_preview_size) = patch.max_preview_size {
            self.max_preview_size = max_preview_size
        }

        self
    }
}

#[cfg(test)]
impl Config {
    pub fn with_max_preview_size(mut self, max_preview_size: Byte) -> Self {
        self.max_preview_size = max_preview_size;
        self
    }
}

#[derive(Debug, Default, Deserialize)]
#[cfg_attr(test, derive(serde::Serialize))]
struct ConfigPatch {
    pub max_preview_size: Option<Byte>,
}

#[cfg(test)]
mod test {
    use std::io::Write;

    use super::*;

    #[test]
    fn config_patch_test() {
        let config = Config::default();
        let patch = ConfigPatch {
            max_preview_size: None,
        };

        let config = config.patch(patch);
        assert_eq!(config, Config::default());

        let patch = ConfigPatch {
            max_preview_size: Some(Byte::from_u64(123)),
        };
        let config = config.patch(patch);
        assert_eq!(
            config,
            Config::default().with_max_preview_size(Byte::from_u64(123))
        );
    }

    #[test]
    fn config_patch_from_files() {
        setup_file("/tmp/jedit-config-bogus", "bogus");
        let config = Config::default().patch_from_files(&["/tmp/jedit-config-bogus"]);
        assert_eq!(config, Config::default());

        setup_file(
            "/tmp/jedit-config-none",
            &toml::to_string_pretty(&ConfigPatch {
                max_preview_size: None,
            })
            .unwrap(),
        );
        let config = Config::default().patch_from_files(&["/tmp/jedit-config-none"]);
        assert_eq!(config, Config::default());

        setup_file(
            "/tmp/jedit-config-some",
            &toml::to_string_pretty(&ConfigPatch {
                max_preview_size: Some(Byte::from_u64(123)),
            })
            .unwrap(),
        );
        let config = Config::default().patch_from_files(&["/tmp/jedit-config-some"]);
        assert_eq!(
            config,
            Config::default().with_max_preview_size(Byte::from_u64(123))
        );

        setup_file(
            "/tmp/jedit-config-some-2",
            &toml::to_string_pretty(&ConfigPatch {
                max_preview_size: Some(Byte::from_u64(1234)),
            })
            .unwrap(),
        );
        let config = Config::default()
            .patch_from_files(&["/tmp/jedit-config-some", "/tmp/jedit-config-some-2"]);
        assert_eq!(
            config,
            Config::default().with_max_preview_size(Byte::from_u64(1234))
        );
    }

    fn setup_file(file_path: &str, content: &str) {
        let mut file = File::create(file_path).unwrap();
        file.write_all(content.as_bytes()).unwrap();
    }
}
