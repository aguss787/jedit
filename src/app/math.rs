#[derive(Debug)]
#[cfg_attr(test, derive(PartialEq, Clone, Copy))]
pub(crate) enum Op {
    Add(u16),
    Sub(u16),
}

const MAGNITUDE: u16 = 5;

impl Op {
    pub(crate) fn exec(self, num: u16) -> u16 {
        let num = num / MAGNITUDE;
        let num = match self {
            Op::Add(n) => num.saturating_add(n),
            Op::Sub(n) => num.saturating_sub(n),
        };
        num * MAGNITUDE
    }
}
