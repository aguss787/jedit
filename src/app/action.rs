use std::collections::VecDeque;

use crate::container::node::Node;

use super::math::Op;

#[derive(Debug, Clone, Copy)]
#[cfg_attr(test, derive(PartialEq))]
pub enum PreviewNavigationAction {
    Up(u16),
    Down(u16),
    Left,
    Right,
}

impl From<PreviewNavigationAction> for Action {
    fn from(value: PreviewNavigationAction) -> Self {
        NavigationAction::PreviewNavigation(value).into()
    }
}

impl From<PreviewNavigationAction> for WorkSpaceAction {
    fn from(value: PreviewNavigationAction) -> Self {
        NavigationAction::PreviewNavigation(value).into()
    }
}

#[derive(Debug)]
#[cfg_attr(test, derive(PartialEq, Clone, Copy))]
pub enum NavigationAction {
    Up(usize),
    Down(usize),
    Top,
    Bottom,
    Expand,
    Close,
    TogglePreview,
    PreviewNavigation(PreviewNavigationAction),
    PreviewWindowResize(Op),
}

impl From<NavigationAction> for Action {
    fn from(value: NavigationAction) -> Self {
        WorkSpaceAction::from(value).into()
    }
}

impl From<NavigationAction> for WorkSpaceAction {
    fn from(value: NavigationAction) -> Self {
        WorkSpaceAction::Navigation(value)
    }
}

#[derive(Debug)]
#[cfg_attr(test, derive(PartialEq, Clone, Copy))]
pub enum ConfirmAction<T, C = bool> {
    Request(T),
    Confirm(C),
}

impl<T, C> ConfirmAction<T, C> {
    pub(crate) fn action_confirmer<R: Into<Action>>(
        f: impl Fn(ConfirmAction<T, C>) -> R,
    ) -> impl Fn(C) -> Action {
        move |b| f(ConfirmAction::Confirm(b)).into()
    }
}

#[derive(Debug)]
#[cfg_attr(test, derive(PartialEq, Clone))]
pub(crate) enum WorkSpaceAction {
    Navigation(NavigationAction),
    Edit,
    EditError(ConfirmAction<String>),
    Save(ConfirmAction<()>),
    SaveDone,
    ErrorConfirmed,
    Load { node: Node, is_edit: bool },
    Rename(ConfirmAction<(), Option<String>>),
}

impl From<WorkSpaceAction> for Action {
    fn from(value: WorkSpaceAction) -> Self {
        Self::Workspace(value)
    }
}

#[derive(Debug)]
#[cfg_attr(test, derive(PartialEq))]
pub enum EditJobAction {
    Init,
    Open,
}

#[derive(Debug)]
#[cfg_attr(test, derive(PartialEq))]
pub enum JobAction {
    Edit(EditJobAction),
    Save,
}

impl From<JobAction> for Action {
    fn from(value: JobAction) -> Self {
        Self::ExecuteJob(value)
    }
}

#[must_use]
#[derive(Debug)]
#[cfg_attr(test, derive(PartialEq))]
pub(crate) enum Action {
    Exit(ConfirmAction<()>),
    Workspace(WorkSpaceAction),
    ExecuteJob(JobAction),
}

pub struct Actions(VecDeque<Action>);

impl Actions {
    pub fn new() -> Self {
        Self(VecDeque::new())
    }

    pub(crate) fn push(&mut self, action: Action) {
        self.0.push_back(action);
    }

    pub(crate) fn next(&mut self) -> Option<Action> {
        self.0.pop_front()
    }

    #[cfg(test)]
    pub(crate) fn into_vec(self) -> Vec<Action> {
        self.0.into_iter().collect()
    }
}

impl Default for Actions {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn actions_in_order_test() {
        let mut actions = Actions::new();
        actions.push(Action::Workspace(WorkSpaceAction::Edit));
        actions.push(Action::Exit(ConfirmAction::Request(())));
        assert_eq!(
            actions.next(),
            Some(Action::Workspace(WorkSpaceAction::Edit))
        );
        assert_eq!(
            actions.next(),
            Some(Action::Exit(ConfirmAction::Request(())))
        );
        assert_eq!(actions.next(), None);
    }
}
