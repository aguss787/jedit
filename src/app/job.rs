use std::thread::JoinHandle;

use super::action::Action;

#[derive(Debug)]
pub struct Job(JoinHandle<Result<Action, std::io::Error>>);

#[cfg(test)]
impl PartialEq for Job {
    fn eq(&self, _other: &Self) -> bool {
        false
    }
}

impl Job {
    pub fn new<F: FnOnce() -> Result<Action, std::io::Error> + Sync + Send + 'static>(
        f: F,
    ) -> Self {
        Self(std::thread::spawn(f))
    }

    pub fn is_done(&self) -> bool {
        self.0.is_finished()
    }

    pub fn action(self) -> Result<Action, std::io::Error> {
        self.0.join().map_err(|err| {
            std::io::Error::new(std::io::ErrorKind::UnexpectedEof, format!("{err:?}"))
        })?
    }
}
