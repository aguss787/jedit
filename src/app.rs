mod action;
mod component;
mod config;
mod job;
mod math;

use std::{
    fs::File,
    io::{Write, stdout},
    process::Command,
    time::Duration,
};

use action::{
    Action, Actions, ConfirmAction, EditJobAction, JobAction, NavigationAction, WorkSpaceAction,
};
use component::workspace::{WorkSpace, WorkSpaceState};
use config::Config;
use crossterm::{
    ExecutableCommand,
    event::{self, Event, KeyCode},
    terminal::{EnterAlternateScreen, LeaveAlternateScreen, disable_raw_mode, enable_raw_mode},
};
use job::Job;
use ratatui::{DefaultTerminal, Frame};

use crate::{container::node::Node, error::LoadError};

struct GlobalState {
    exit: bool,
}

pub struct CliApp {
    state: GlobalState,
    worktree_state: WorkSpaceState,
    worktree: WorkSpace,
    output_file_name: String,
    jobs: Vec<Job>,
}

impl CliApp {
    pub fn new(input_file_name: String, output_file_name: String) -> std::io::Result<Self> {
        let initial_load_job = Job::new(move || {
            let file = File::open(&input_file_name)?;
            let file_root = Node::load(file).map_err(|error| {
                std::io::Error::new(std::io::ErrorKind::InvalidData, error.to_string())
            })?;

            Ok(WorkSpaceAction::Load(file_root).into())
        });

        let mut cli_app = Self {
            worktree: WorkSpace::new(Node::null(), Config::load()),
            worktree_state: WorkSpaceState::default(),
            state: GlobalState { exit: false },
            output_file_name,
            jobs: vec![initial_load_job],
        };
        cli_app.worktree.decrease_edit_cntr();
        Ok(cli_app)
    }

    pub fn run(&mut self) -> std::io::Result<()> {
        let mut terminal = Terminal::new();

        self.worktree.handle_action(
            &mut self.worktree_state,
            &mut Actions::new(),
            NavigationAction::TogglePreview.into(),
        )?;

        while !self.state.exit {
            terminal.0.draw(|frame| self.draw(frame))?;
            self.handle_event(&mut terminal)?;
        }

        Ok(())
    }

    fn draw(&mut self, frame: &mut Frame) {
        frame.render_stateful_widget(&self.worktree, frame.area(), &mut self.worktree_state);
    }

    fn handle_event(&mut self, terminal: &mut Terminal) -> std::io::Result<()> {
        let mut actions = Actions::new();
        if event::poll(FRAME_TIME)? {
            let event = event::read()?;
            if global_exit_handler(&event) {
                self.state.exit = true;
                return Ok(());
            }

            self.worktree.handle_event(&mut actions, event);
        }

        let mut jobs = Vec::new();
        std::mem::swap(&mut jobs, &mut self.jobs);
        jobs.into_iter()
            .filter_map(|job| {
                if job.is_done() {
                    Some(job.action())
                } else {
                    self.jobs.push(job);
                    None
                }
            })
            .collect::<Result<Vec<_>, _>>()?
            .into_iter()
            .for_each(|action| actions.push(action));

        while let Some(action) = actions.next() {
            match action {
                Action::Exit(confirm_action) => {
                    self.state.exit = self.worktree.maybe_exit(confirm_action);
                    return Ok(());
                }
                Action::Workspace(workspace_action) => self.worktree.handle_action(
                    &mut self.worktree_state,
                    &mut actions,
                    workspace_action,
                )?,
                Action::ExecuteJob(job) => {
                    if let Some(job) = self.execute_job(terminal, job)? {
                        self.jobs.push(job);
                    }
                }
            }
        }

        self.worktree.set_loading(!self.jobs.is_empty());
        Ok(())
    }

    fn execute_job(&self, terminal: &mut Terminal, job: JobAction) -> std::io::Result<Option<Job>> {
        let job = match job {
            JobAction::Edit(EditJobAction::Init) => {
                let Some(node) = self.worktree.selected_node(&self.worktree_state) else {
                    return Ok(None);
                };
                let node = NodeJob(node);
                Job::new(move || {
                    let mut file = File::create(EDITOR_BUFFER)?;
                    let _ = &node;
                    let node = unsafe { node.0.as_ref().expect("invalid pointer to node") };
                    let content = node
                        .to_string_pretty()
                        .expect("invalid internal representation");
                    file.write_all(content.as_bytes())?;
                    Ok(JobAction::Edit(EditJobAction::Open).into())
                })
            }
            JobAction::Edit(EditJobAction::Open) => {
                terminal.run_editor(EDITOR_BUFFER)?;
                Job::new(|| {
                    let file = File::open(EDITOR_BUFFER)?;

                    match Node::load(file) {
                        Err(LoadError::IO(error)) => Err(error),
                        Err(LoadError::SerdeJson(error)) => Ok(WorkSpaceAction::EditError(
                            ConfirmAction::Request(error.to_string()),
                        )
                        .into()),
                        Err(LoadError::DeserializationError(error)) => Ok(
                            WorkSpaceAction::EditError(ConfirmAction::Request(error.to_string()))
                                .into(),
                        ),
                        Ok(node) => Ok(WorkSpaceAction::Load(node).into()),
                    }
                })
            }
            JobAction::Save => {
                let mut output_file = File::create(&self.output_file_name)?;
                let content: *const Node = self.worktree.file_root();
                let content = NodeJob(content);
                Job::new(move || {
                    let _ = &content;
                    let content =
                        unsafe { content.0.as_ref().expect("invalid pointer to content") };
                    output_file.write_all(
                        content
                            .to_string_pretty()
                            .expect("invalid internal representation")
                            .as_bytes(),
                    )?;
                    Ok(WorkSpaceAction::SaveDone.into())
                })
            }
        };

        Ok(Some(job))
    }
}

struct NodeJob(*const Node);
unsafe impl Send for NodeJob {}
unsafe impl Sync for NodeJob {}

fn global_exit_handler(event: &Event) -> bool {
    let Some(key_event) = event.as_key_event() else {
        return false;
    };

    if !key_event.is_press() {
        return false;
    }

    key_event.code == KeyCode::F(5)
}

pub struct Terminal(DefaultTerminal);

impl Terminal {
    fn new() -> Self {
        Self(ratatui::init())
    }

    fn run_editor(&mut self, path: &str) -> std::io::Result<()> {
        let editor = std::env::var("EDITOR")
            .ok()
            .unwrap_or_else(|| String::from("vi"));
        stdout().execute(LeaveAlternateScreen)?;
        disable_raw_mode()?;
        Command::new(&editor).arg(path).status()?;
        stdout().execute(EnterAlternateScreen)?;
        enable_raw_mode()?;
        self.0.clear()?;
        Ok(())
    }
}

impl Drop for Terminal {
    fn drop(&mut self) {
        ratatui::restore();
    }
}

const FRAME_TIME: Duration = Duration::from_millis(16);
const EDITOR_BUFFER: &str = "/tmp/jedit-buffer.json";
