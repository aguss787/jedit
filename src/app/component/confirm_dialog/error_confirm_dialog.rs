use crossterm::event::Event;
use ratatui::{
    layout::Alignment,
    prelude::{Buffer, Rect},
    text::{Line, Text},
    widgets::{Block, Padding, Widget, WidgetRef},
};

use crate::app::{
    action::{Actions, WorkSpaceAction},
    component::popup::BoundedPopUp,
};

use super::ConfirmDialog;

pub struct ErrorConfirmDialog {
    message: Text<'static>,
    title: Option<Line<'static>>,
}

impl ErrorConfirmDialog {
    pub(crate) fn new(message: Text<'static>) -> Self {
        Self {
            message,
            title: None,
        }
    }

    pub(crate) fn title(mut self, title: Line<'static>) -> Self {
        self.title = Some(title);
        self
    }
}

impl ConfirmDialog for ErrorConfirmDialog {
    fn handle_event(&self, actions: &mut Actions, event: Event) {
        if !event.is_key_press() {
            return;
        };

        actions.push(WorkSpaceAction::ErrorConfirmed.into());
    }
}

impl WidgetRef for ErrorConfirmDialog {
    fn render_ref(&self, area: Rect, buf: &mut Buffer) {
        let title = self.title.clone().unwrap_or_else(|| "Error!".into());
        let title_width = title.width() as u16;
        let block = Block::bordered()
            .padding(Padding::symmetric(1, 1))
            .title_top(title)
            .title_bottom(Line::from("Press any key"))
            .title_alignment(Alignment::Center);

        BoundedPopUp::new(block, self.message.clone())
            .min_width(title_width.max(20))
            .render(area, buf);
    }
}

#[cfg(test)]
mod test {
    use insta::assert_snapshot;

    use crate::app::component::test_render::render_to_string;

    use super::*;

    #[test]
    fn render_default_test() {
        assert_snapshot!(render_to_string(&ErrorConfirmDialog::new(
            "short error!".into()
        )));
    }

    #[test]
    fn render_title_test() {
        assert_snapshot!(render_to_string(
            &ErrorConfirmDialog::new("short error!".into()).title(Line::from("Short title"))
        ));
    }

    #[test]
    fn render_long_title_test() {
        assert_snapshot!(render_to_string(
            &ErrorConfirmDialog::new("short error!".into())
                .title(Line::from("This is a very long title"))
        ));
    }
}
