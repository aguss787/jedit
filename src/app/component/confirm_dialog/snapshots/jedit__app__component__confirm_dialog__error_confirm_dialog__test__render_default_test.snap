---
source: src/app/component/confirm_dialog/error_confirm_dialog.rs
expression: "render_to_string(&ErrorConfirmDialog::new(\"short error!\".into()))"
---
"                                                                                "
"                                                                                "
"                                                                                "
"                                                                                "
"                                                                                "
"                                                                                "
"                                                                                "
"                                                                                "
"                            ┌────────Error!────────┐                            "
"                            │                      │                            "
"                            │ short error!         │                            "
"                            │                      │                            "
"                            └────Press any key─────┘                            "
"                                                                                "
"                                                                                "
"                                                                                "
"                                                                                "
"                                                                                "
"                                                                                "
"                                                                                "
